evalml/tests/pipeline_tests/test_component_graph.py

```diff
@@ -0,0 +1,332 @@
+from unittest.mock import patch
+
+import numpy as np
+import pandas as pd
+import pytest
+
+from evalml.exceptions import MissingComponentError
+from evalml.pipelines import ComponentGraph
+from evalml.pipelines.components import (
+    ElasticNetClassifier,
+    Estimator,
+    Imputer,
+    LogisticRegressionClassifier,
+    OneHotEncoder,
+    RandomForestClassifier
+)
+
+
+@pytest.fixture
+def example_graph():
+    graph = {'Imputer': [Imputer],
+             'OneHot_RandomForest': [OneHotEncoder, 'Imputer.x'],
+             'OneHot_ElasticNet': [OneHotEncoder, 'Imputer.x'],
+             'Random Forest': [RandomForestClassifier, 'OneHot_RandomForest.x'],
+             'Elastic Net': [ElasticNetClassifier, 'OneHot_ElasticNet.x'],
+             'Logistic Regression': [LogisticRegressionClassifier, 'Random Forest', 'Elastic Net']}
+    return graph
+
+
+def test_init(example_graph):
+    comp_graph = ComponentGraph()
+    assert len(comp_graph.component_dict) == 0
+
+    graph = example_graph
+    comp_graph = ComponentGraph(graph)
+    assert len(comp_graph.component_dict) == 6
+
+    expected_order = ['Imputer', 'OneHot_ElasticNet', 'Elastic Net', 'OneHot_RandomForest', 'Random Forest', 'Logistic Regression']
+    assert comp_graph.compute_order == expected_order
+
+    invalid_graph = {'Imputer': [Imputer], 'OHE': OneHotEncoder}
+    with pytest.raises(ValueError, match='All component information should be passed in as a list'):
+        ComponentGraph(invalid_graph)
```

Comment:
Great idea, I added tests for all these cases! To answer your questions:
- The implementation does support string names for components, since the same `handle_component_class` method is called
- NetworkX catches any circular dependencies, so I've added an extra catch in `_recompute_order` so we throw that error ourselves
- The current implementation did not detect graphs that aren't fully connected, so I've added a check for that in `_recompute_order` as well